src/main/java/io/jenkins/jenkins/plugins/okhttp/api/JenkinsOkHttpClient.java

```diff
@@ -0,0 +1,91 @@
+package io.jenkins.jenkins.plugins.okhttp.api;
+
+import hudson.ProxyConfiguration;
+import jenkins.model.Jenkins;
+import okhttp3.Authenticator;
+import okhttp3.Credentials;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.Route;
+
+import javax.net.ssl.SSLSocketFactory;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.SocketAddress;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+
+/**
+ * This class allows to take into consideration if Jenkins is running behind a proxy or not and return a proper {@link OkHttpClient}
+ * client.
+ */
+public class JenkinsOkHttpClient {
+
+    private static final Logger LOGGER = Logger.getLogger(JenkinsOkHttpClient.class.getName());
+
+    private JenkinsOkHttpClient() {
+    }
+
+    /**
+     * Generates a new builder for the given base client.
+     * It applies the current {@link Jenkins#proxy} configuration.
+     *
+     * @return a builder to configure the client
+     */
+    public static OkHttpClient.Builder newClientBuilder(OkHttpClient httpClient) {
+        OkHttpClient.Builder reBuild = httpClient.newBuilder();
+        if (Jenkins.get().proxy != null) {
+            final ProxyConfiguration proxy = Jenkins.get().proxy;
+            Proxy proxyServer = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxy.name, proxy.port));
+            if (proxy.getUserName() != null) {
+                Authenticator proxyAuthenticator = new Authenticator() {
+                    @Override public Request authenticate(Route route, Response response) throws IOException {
+                        String credential = Credentials.basic(proxy.getUserName(), proxy.getPassword());
+                        return response.request().newBuilder()
+                                       .header("Proxy-Authorization", credential)
+                                       .build();
+                    }
+                };
+                reBuild.proxyAuthenticator(proxyAuthenticator);
+            }
+
+            // TODO: https://github.com/square/okhttp/issues/3787
+            // not sure if this is correct, needs manual testing
+            if (proxy.name.startsWith("https")) {
```

Comment:
what about proxy name `httpsserverwithoutssl.mydomain.com` :-o

please look for startswith `https://` and as schemes are case insensitive check without case sensitivity :)

Reply:
Dummy questions: can `proxy.name` omit the scheme? Can it be an IP (with or without the scheme)? I had the idea of trying to convert it to an URL and check the protocol there, and if it's not a valid URL, previous questions are valid.

Reply:
I've never seen it with a scheme before - normally is it is just a server name (or IP address).
given how it is used in https://github.com/jenkinsci/jenkins/blob/6adc9c0519073b6ab72c0c3d5780d0e2ec21b294/core/src/main/java/hudson/ProxyConfiguration.java#L266 I would expect it to be a hostname or IP address and never have a scheme or anything else.